sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/ReleaseDownloadListener.java

```diff
@@ -0,0 +1,196 @@
+
+            /*
+             * This start call triggers strict mode in UI thread so it
+             * needs to be done here without synchronizing
+             * (not to block methods waiting on synchronized on UI thread)
+             * so yes we could launch install and SDK being disabled.
+             *
+             * This corner case cannot be avoided without triggering
+             * strict mode exception.
+             */
+            AppCenterLog.info(LOG_TAG, "Show install UI for " + localUri);
+            mContext.startActivity(intent);
+            Distribute distribute = Distribute.getInstance();
+            if (mReleaseDetails.isMandatoryUpdate()) {
+                distribute.setInstalling(mReleaseDetails);
+            } else {
+                distribute.completeWorkflow(mReleaseDetails);
+            }
+            storeReleaseDetails(mReleaseDetails);
+        }
+        return true;
+    }
+
+    @Override
+    public void onError(@NonNull String errorMessage) {
+        AppCenterLog.error(LOG_TAG, errorMessage);
+        Toast.makeText(mContext, R.string.appcenter_distribute_downloading_error, Toast.LENGTH_SHORT).show();
+        Distribute.getInstance().completeWorkflow(mReleaseDetails);
+    }
+
+    /**
+     * Show download progress. Only used for mandatory updates.
+     */
+    @SuppressWarnings({"deprecation", "RedundantSuppression"})
+    android.app.ProgressDialog showDownloadProgress(Activity foregroundActivity) {
+        if (!mReleaseDetails.isMandatoryUpdate()) {
+            return null;
+        }
+        mProgressDialog = new android.app.ProgressDialog(foregroundActivity);
+        mProgressDialog.setTitle(R.string.appcenter_distribute_downloading_update);
+        mProgressDialog.setCancelable(false);
+        mProgressDialog.setProgressStyle(android.app.ProgressDialog.STYLE_HORIZONTAL);
+        mProgressDialog.setIndeterminate(true);
+        mProgressDialog.setProgressNumberFormat(null);
+        mProgressDialog.setProgressPercentFormat(null);
+        return mProgressDialog;
+    }
+
+    /**
+     * Hide progress dialog and stop updating. Only used for mandatory updates.
+     */
+    @SuppressWarnings({"deprecation", "RedundantSuppression"})
+    synchronized void hideProgressDialog() {
+        if (mProgressDialog != null) {
+            final android.app.ProgressDialog progressDialog = mProgressDialog;
+            mProgressDialog = null;
+
+            /* This can be called from background check download task. */
+            HandlerUtils.runOnUiThread(new Runnable() {
+
+                @Override
+                public void run() {
+                    progressDialog.hide();
+                }
+            });
+            HandlerUtils.getMainHandler().removeCallbacksAndMessages(HANDLER_TOKEN_CHECK_PROGRESS);
+        }
+    }
+
+    private static void storeReleaseDetails(@NonNull ReleaseDetails releaseDetails) {
```

Comment:
This was a bug also in develop, but we need to move that to Distribute so that it checks for state change and synchronize (like setInstalling was there for synchronize/state check reasons).

Reply:
Can file a bug instead as this should also be the case in develop.